core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java

```diff
@@ -32,74 +32,107 @@
+        List<TypeOfCancer> typeOfCancerList = parseCancerTypesFromFile(file);
+        SpringUtil.initDataSource();
+        if (clobber) {
+            ProgressMonitor.setCurrentMessage("Deleting all previous cancer types...");
+            DaoTypeOfCancer.deleteAllRecords(); //TODO - remove this option - foreign key constraints may mean large cascade effects (possibly the deletion of all studies) - instead, change the option to 'deleteTypeOfCancerIfNotPresent' and add a loop through existing typeOfCancer records, removing those which are not in the parsed typeOfCancerList
+        }
+        writeRecordsToDatabase(typeOfCancerList, clobber);
+        ProgressMonitor.setCurrentMessage("Done.");
+        ConsoleUtil.showMessages();
+    }
+
+    private static List<TypeOfCancer> parseCancerTypesFromFile(File file) throws IOException {
+        ProgressMonitor.setCurrentMessage(String.format("Reading cancer types from file '%s'...", file.getPath()));
+        List<TypeOfCancer> typeOfCancerList = new ArrayList<TypeOfCancer>();
         Scanner scanner = new Scanner(file);
-        int numNewCancerTypes = 0;
-        
-        while(scanner.hasNextLine()) {
-            String[] tokens = scanner.nextLine().split("\t", -1);
-            if (tokens.length != 4) {
-                throw new IOException(
-                    "Cancer type file '" + file.getPath() +
-                    "' is not a five-column tab-delimited file");
-            }
-            
-            String typeOfCancerId = tokens[0].trim();
-            //if not clobber, then existing cancer types should be skipped:
-            if (!clobber && DaoTypeOfCancer.getTypeOfCancerById(typeOfCancerId.toLowerCase()) != null ) {
-            	ProgressMonitor.logWarning("Cancer type with id '" + typeOfCancerId + "' already exists. Skipping.");
+        while (scanner.hasNextLine()) {
+            String nextLine = scanner.nextLine();
+            String[] fields = nextLine.split("\t", -1);
+            throwExceptionIfColumnCountIsWrong(file, nextLine, fields, EXPECTED_DATAFILE_COLUMN_COUNT);
+            TypeOfCancer typeOfCancer = new TypeOfCancer();
+            String typeOfCancerId = fields[0].trim();
+            typeOfCancer.setTypeOfCancerId(typeOfCancerId.toLowerCase());
+            typeOfCancer.setName(fields[1].trim());
+            typeOfCancer.setDedicatedColor(fields[2].trim());
+            typeOfCancer.setShortName(typeOfCancerId);
+            typeOfCancer.setParentTypeOfCancerId(fields[3].trim().toLowerCase());
+            typeOfCancerList.add(typeOfCancer);
+        }
+        return typeOfCancerList;
+    }
+
+    private static void throwExceptionIfColumnCountIsWrong(File file, String nextLine, String[] fields, int expectedDatafileColumnCount) throws IOException {
+        if (fields.length != expectedDatafileColumnCount) {
+            String msg = String.format("Cancer type file '%s' contains a line which is not %d-column tab-delimited." +
```

Comment:
the String.format to create String msg has 4 parameters, but only 3 have been passed. It requires expectedDatafileColumnCount also as a parameter.